exes/mqtt-broadcaster/inc/impl.hpp

```diff
@@ -0,0 +1,129 @@
+#pragma once
+
+#include <sparkplug_b/sparkplug_b.pb.h>
+
+using org::eclipse::tahu::protobuf::Payload;
+using org::eclipse::tahu::protobuf::Payload_Metric;
+
+namespace tfc::mqtt::impl {
+
+struct signal_data {
+  tfc::ipc_ruler::signal information;
+  tfc::ipc::details::any_recv receiver;
+  std::optional<std::any> current_value;
+};
+
+// Create a timestamp in milliseconds from 1 Jan 1970 according to spec
+static auto timestamp_milliseconds() -> std::chrono::milliseconds {
+  using std::chrono::duration_cast;
+  using std::chrono::milliseconds;
+  using std::chrono::system_clock;
+  return duration_cast<milliseconds>(system_clock::now().time_since_epoch());
+}
+
+/// This function converts tfc types to Spark Plug B types
+/// More information can be found (page 76) in the spec under section 6.4.16 data types:
+/// https://sparkplug.eclipse.org/specification/version/3.0/documents/sparkplug-specification-3.0.0.pdf
+auto type_enum_convert(tfc::ipc::details::type_e type) -> uint32_t {
+  using enum tfc::ipc::details::type_e;
+  switch (type) {
+    case unknown:
+      return 0;
```

Comment:
can this be indexed into an enum instead of literal number?

Reply:
if so the text above can be removed

Reply:
The number is used to set the datatype according to the SparkPlugB specification

```cpp
variable_metric.set_datatype(impl::type_enum_convert(signal_data.information.type));
```